repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/AbstractKeystoneV2Filter.scala

```diff
@@ -0,0 +1,146 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.keystonev2
+
+import javax.servlet._
+import javax.servlet.http.HttpServletResponse._
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+import com.rackspace.httpdelegation.HttpDelegationManager
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.utils.http.{IdentityStatus, OpenStackServiceHeader}
+import org.openrepose.commons.utils.servlet.http.ResponseMode.{MUTABLE, PASSTHROUGH}
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper}
+import org.openrepose.core.filter.AbstractConfiguredFilter
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.keystonev2.config._
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
+
+abstract class AbstractKeystoneV2Filter[T <: KeystoneV2Config: ClassTag](configurationService: ConfigurationService)
+  extends AbstractConfiguredFilter[T](configurationService)
+    with HttpDelegationManager
+    with LazyLogging {
+
+  import AbstractKeystoneV2Filter._
+
+  val handleFailures: PartialFunction[Try[Unit.type], KeystoneV2Result]
+
+  def doAuth(request: HttpServletRequestWrapper): Try[Unit.type]
+
+  def doWork(servletRequest: HttpServletRequest, servletResponse: HttpServletResponse, chain: FilterChain): Unit = {
+    /**
+      * STATIC REFERENCE TO CONFIG
+      */
+    val config = configuration
+
+    /**
+      * DECLARE COMMON VALUES
+      */
+    lazy val request = new HttpServletRequestWrapper(servletRequest)
+    lazy val response = new HttpServletResponseWrapper(servletResponse, MUTABLE, PASSTHROUGH)
+
+    def isWhitelisted(requestUri: String): Boolean = {
+      logger.trace("Comparing request URI to whitelisted URIs")
+
+      val whiteListUris: List[String] = config.getWhiteList.getUriRegex.asScala.toList
+
+      whiteListUris exists { pattern =>
+        logger.debug(s"checking $requestUri against $pattern")
+        requestUri.matches(pattern)
+      }
+    }
+
+    /**
+      * BEGIN PROCESSING
+      */
+    logger.debug("Keystone v2 filter processing request...")
+
+    val filterResult =
+      if (isWhitelisted(request.getRequestURI)) {
+        Pass
+      } else {
+        val authResult = doAuth(request)
+        handleFailures.applyOrElse(authResult, (t: Try[Unit.type]) => t match {
+          case Success(_) => Pass
+          case Failure(e) => Reject(SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
+        })
```

Comment:
This could also be written as:

```
        handleFailures.applyOrElse(authResult, (_: Try[Unit.type]) match {
          case Success(_) => Pass
          case Failure(e) => Reject(SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
        })
```

Unfortunately, the argument types of an anonymous function must be fully known, so that's as simple as I could get it.